From 1b02fe3ab446613cadcfac08d611b90becb02189 Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.org>
Date: Wed, 18 Apr 2018 17:04:23 +0100
Subject: [PATCH 277/454] overlays: Add 'combine' option to i2c overlays

The i2c0-bcm2708 and i2c1-bcm2708 overlays allow the I2C pin usage to
be changed. The names also suggest they revert to the old i2c_bcm2708
driver, but they don't. The newer i2c_bcm2835 driver forces
transactions to be combined where possible, but this breaks some
devices.

Add an option to disable transaction combining, which is currently
implemented by reverting to the old driver.

See: https://github.com/raspberrypi/firmware/issues/828

Signed-off-by: Phil Elwell <phil@raspberrypi.org>
---
 arch/arm/boot/dts/overlays/README                   | 12 ++++++++++--
 arch/arm/boot/dts/overlays/i2c0-bcm2708-overlay.dts |  8 ++++++++
 arch/arm/boot/dts/overlays/i2c1-bcm2708-overlay.dts |  9 +++++++++
 3 files changed, 27 insertions(+), 2 deletions(-)

--- a/arch/arm/boot/dts/overlays/README
+++ b/arch/arm/boot/dts/overlays/README
@@ -865,8 +865,9 @@ Params: addr                    Set the
 
 
 Name:   i2c0-bcm2708
-Info:   Enable the i2c_bcm2708 driver for the i2c0 bus. Not all pin combinations
-        are usable on all platforms.
+Info:   Change i2c0 pin usage. Not all pin combinations are usable on all
+        platforms - platforms other then Compute Modules can only use this
+        to disable transaction combining.
 Load:   dtoverlay=i2c0-bcm2708,<param>=<val>
 Params: sda0_pin                GPIO pin for SDA0 (deprecated - use pins_*)
         scl0_pin                GPIO pin for SCL0 (deprecated - use pins_*)
@@ -874,15 +875,22 @@ Params: sda0_pin                GPIO pin
         pins_28_29              Use pins 28 and 29
         pins_44_45              Use pins 44 and 45
         pins_46_47              Use pins 46 and 47
+        combine                 Allow transactions to be combined (default
+                                "yes")
 
 
 Name:   i2c1-bcm2708
+Info:   Change i2c1 pin usage. Not all pin combinations are usable on all
+        platforms - platforms other then Compute Modules can only use this
+        to disable transaction combining.
 Info:   Enable the i2c_bcm2708 driver for the i2c1 bus
 Load:   dtoverlay=i2c1-bcm2708,<param>=<val>
 Params: sda1_pin                GPIO pin for SDA1 (2 or 44 - default 2)
         scl1_pin                GPIO pin for SCL1 (3 or 45 - default 3)
         pin_func                Alternative pin function (4 (alt0), 6 (alt2) -
                                 default 4)
+        combine                 Allow transactions to be combined (default
+                                "yes")
 
 
 Name:   i2s-gpio28-31
--- a/arch/arm/boot/dts/overlays/i2c0-bcm2708-overlay.dts
+++ b/arch/arm/boot/dts/overlays/i2c0-bcm2708-overlay.dts
@@ -50,6 +50,13 @@
       };
    };
 
+   fragment@5 {
+      target = <&i2c0>;
+      __dormant__ {
+         compatible = "brcm,bcm2708-i2c";
+      };
+   };
+
    __overrides__ {
       sda0_pin = <&frag1>,"brcm,pins:0";
       scl0_pin = <&frag1>,"brcm,pins:4";
@@ -57,5 +64,6 @@
       pins_28_29 = <0>,"-1+2-3-4";
       pins_44_45 = <0>,"-1-2+3-4";
       pins_46_47 = <0>,"-1-2-3+4";
+      combine = <0>, "!5";
    };
 };
--- a/arch/arm/boot/dts/overlays/i2c1-bcm2708-overlay.dts
+++ b/arch/arm/boot/dts/overlays/i2c1-bcm2708-overlay.dts
@@ -26,9 +26,18 @@
          brcm,function = <4>; /* alt 0 */
       };
    };
+
+   fragment@2 {
+      target = <&i2c1>;
+      __dormant__ {
+         compatible = "brcm,bcm2708-i2c";
+      };
+   };
+
    __overrides__ {
       sda1_pin = <&pins>,"brcm,pins:0";
       scl1_pin = <&pins>,"brcm,pins:4";
       pin_func = <&pins>,"brcm,function:0";
+      combine = <0>, "!2";
    };
 };
